### PR TITLE
Add SimplexNoise and NoiseTexture as new resources

### PR DESCRIPTION
**Pull request by JFonS**
SimplexNoise can be used to generate parameterized fractal noise based on Open Simplex, intended to be used in GDsrcipt.

NoiseTexture uses SimplexNoise to generate noise textures for using in
shaders/visual effects.

Documentation is missing, I will make a separate PR for that later on.

Edit: added option to use noise as a normalmap,

noise0
noise1

### CODE REVIEW

**Comment by Zylann** [comment]
Should the class be named `SimplexNoise` even though the algorithm is actually `OpenSimplex`?

**Comment by Megalomaniak** [comment]
Yes, I think `SimplexNoise` is better. A new user might not care for nor need to understand the difference so long as they are just familiar enough with noise types to know that it is a Perlin type noise function.

**Comment by Xrayez** [comment]
If the concern is the patent issue with simplex noise then based on my estimation it should expire on 2021 so we can be strategical about naming convention.. Besides, OpenSimplex implementation could be replaced with Simplex in the future.

File: thirdparty/open-simplex-noise/open-simplex-noise.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+#else 
+	/* ANSI C doesn't have inline or stdint.h. */
+	#define INLINE
+#endif
+
+#ifdef __cplusplus
+	extern "C" {
+#endif
+
+// -- GODOT start --
```
Comment (akien-mga) [inline thirdparty/open-simplex-noise/open-simplex-noise.h]: Why are those changes needed? It should be documented (or ideally avoided if we can do without).

File: thirdparty/README.md [hunk 2]
```diff
@@ -393,6 +393,23 @@ Files extracted from the upstream source:
 - All .h files in `src/`
 - LICENSE.txt
 
+##open-simplex-noise
+
+- Upstream: https://github.com/smcameron/open-simplex-noise-in-c
+- Version: git (0d555e7, 2018)
```
Comment (akien-mga) [inline thirdparty/README.md]: That's 2015 :)

File: thirdparty/README.md [hunk 3]
```diff
@@ -393,6 +393,23 @@ Files extracted from the upstream source:
 - All .h files in `src/`
 - LICENSE.txt
 
+##open-simplex-noise
```
Comment (akien-mga) [inline thirdparty/README.md]: There should be a space after `##`, and two lines between sections.

That being said, since it's only two files, I think it would be better to just put them in `thirdparty/misc` (and document them in the relevant section). You can likely drop the README and maybe name the license file `open-simplex-noise-LICENSE`.

File: thirdparty/README.md [hunk 4]
```diff
@@ -393,6 +393,23 @@ Files extracted from the upstream source:
...
+
+- Upstream: https://github.com/smcameron/open-simplex-noise-in-c
+- Version: git (0d555e7, 2018)
+- License: The Unlicense
```
Comment (akien-mga) [inline thirdparty/README.md]: The short identifier is just "Unlicense": https://spdx.org/licenses/Unlicense.html

File: thirdparty/open-simplex-noise/open-simplex-noise.h (hunk 1 above)
Comment (akien-mga) [reply thirdparty/open-simplex-noise/open-simplex-noise.h]: Ah I saw the comment in the README.md. It would be nice to have a patch file like `./thirdparty/minizip/godot-zlib-1.2.4-minizip-seek.patch` so that it can be easily reapplied when syncing with upstream.

File: modules/opensimplex/config.py [hunk 5]
```diff
@@ -0,0 +1,11 @@
+
```
Comment (akien-mga) [inline modules/opensimplex/config.py]: Please remove leading and trailing newlines (see other `config.py` files).

You could also add definitions to get the documentation included in the module itself:
```
def get_doc_classes():
    return [
        "NoiseTexture",
        "SimplexNoise",
    ]

def get_doc_path():
    return "doc_classes"
```

File: modules/opensimplex/noise.cpp [hunk 6]
```diff
@@ -0,0 +1,226 @@
+#include "noise.h"
```
Comment (akien-mga) [inline modules/opensimplex/noise.cpp]: Please add Godot's copyright header.

I'd also suggest to rename `noise.{cpp,h}` to `simplex_noise.{cpp,h}` to match the class name.

File: modules/opensimplex/noise.cpp [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+#include "core_string_names.h"
```
Comment (akien-mga) [inline modules/opensimplex/noise.cpp]: This should be `core/core_string_names.h`, I removed `core/` from the include path. Please also add a line between `noise.h` (or `simplex_noise.h`) and other includes.

File: modules/opensimplex/noise.h [hunk 8]
```diff
@@ -0,0 +1,62 @@
+#ifndef OPENSIMPLEX_NOISE_H
```
Comment (akien-mga) [inline modules/opensimplex/noise.h]: Copyright header needed here too. The guard should likely be renamed to match the class and filename (so `SIMPLEX_NOISE_H` once renamed to `simplex_noise.h`).

File: modules/opensimplex/noise.h [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+#define OPENSIMPLEX_NOISE_H
+
+#include "image.h"
+#include "open-simplex-noise.h"
+#include "scene/resources/texture.h"
+#include <reference.h>
```
Comment (akien-mga) [inline modules/opensimplex/noise.h]: Godot headers are included with `""`, `<>` is for system/thirdparty headers. And `core/` includes should be absolute. So this should be:
```cpp
#include "core/image.h"
#include "core/reference.h"
#include "scene/resources/texture.h"

#include <open-simplex-noise.h>
```

File: modules/opensimplex/noise_texture.cpp [hunk 10]
```diff
@@ -0,0 +1,213 @@
+#include "noise_texture.h"
+#include "core_string_names.h"
```
Comment (akien-mga) [inline modules/opensimplex/noise_texture.cpp]: Same here, add `core/` and a newline.

File: modules/opensimplex/noise_texture.h [hunk 11]
```diff
@@ -0,0 +1,73 @@
+#ifndef NOISE_TEXTURE_H
+#define NOISE__TEXTURE_H
+
+#include "editor/editor_node.h"
+#include "editor/editor_plugin.h"
+#include "editor/property_editor.h"
+#include "image.h"
+#include "noise.h"
+#include <reference.h>
```
Comment (akien-mga) [inline modules/opensimplex/noise_texture.h]: Same comment as in `noise.h`.

File: modules/opensimplex/doc_classes/SimplexNoise.xml [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+	<methods>
+		<method name="get_image">
+			<return type="Image">
+			</return>
+			<argument index="0" name="width" type="int">
+			</argument>
+			<argument index="1" name="height" type="int">
+			</argument>
+			<description>
+				Generate a noise image with the requested [code]width[\code] and [code]height[\code], based on the current noise parameters.
```
Comment (akien-mga) [inline modules/opensimplex/doc_classes/SimplexNoise.xml]: You got those slashes backwards :)

File: modules/opensimplex/doc_classes/SimplexNoise.xml [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+			</description>
+		</method>
+		<method name="get_seamless_image">
+			<return type="Image">
+			</return>
+			<argument index="0" name="size" type="int">
+			</argument>
+			<description>
+				Generate a tileable noise image, based on the current noise parameters. 
+				Generated seamless images are always square ([code]size[\code]x[code]size[\code]).
```
Comment (akien-mga) [inline modules/opensimplex/doc_classes/SimplexNoise.xml]: Here too.

File: modules/opensimplex/SCsub [hunk 14]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env python
+
+Import('env')
+
+thirdparty_dir = '#thirdparty/misc/'
+thirdparty_source = thirdparty_dir + 'open-simplex-noise.c'
+env.Append(CPPPATH=thirdparty_dir)
```
Comment (akien-mga) [inline modules/opensimplex/SCsub]: Since it's only one file, and not likely to be packaged by Linux distros and thus unbundled for the time being, I'd suggest not to include `thirdparty/misc` to the include path, and instead include `thirdparty/misc/open-simplex-noise.h` directly in the relevant files.

This saves us an extra 18 characters on every single compilation command line, which is nice to avoid logs growing to crazy sizes and Windows aborting the build due to too long instructions.

File: modules/opensimplex/register_types.cpp [hunk 15]
```diff
@@ -0,0 +1,12 @@
+#include "register_types.h"
```
Comment (akien-mga) [inline modules/opensimplex/register_types.cpp]: Those take copyright headers too for consistency, even if trivial.

File: modules/opensimplex/register_types.h [hunk 16]
```diff
@@ -0,0 +1,2 @@
+void register_opensimplex_types();
```
Comment (akien-mga) [inline modules/opensimplex/register_types.h]: Here too.

File: modules/opensimplex/simplex_noise.h [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+/*************************************************************************/
+
+#ifndef SIMPLEX_NOISE_H
+#define SIMPLEX_NOISE_H
+
+#include "core/image.h"
+#include "core/reference.h"
+#include "scene/resources/texture.h"
+
+#include <open-simplex-noise.h>
```
Comment (akien-mga) [inline modules/opensimplex/simplex_noise.h]: If you do change it to `thirdparty/misc/open-simplex-noise.h` as I now suggested, it should go back to using double quotes as it's no longer a "system" include.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks a ton! :)

**Comment by avencherus** [comment]
Fantastic!  XD  Going to try it out soon.

**Comment by PetePete1984** [comment]
Before it's finalized in a release build: the correct spelling would be "persistence", not "persistance"

**Comment by smcameron** [comment]
Hello.  

> If the concern is the patent issue with simplex noise then based on my estimation it should expire on 2021 so we can be strategical about naming convention.. Besides, OpenSimplex implementation could be replaced with Simplex in the future.

An argument against changing the underlying noise function in the future would be that this would change the behavior of any programs relying on it for e.g. deterministic procedural generation, which for some programs (e.g. something along the lines of No Man's Sky) this would amount to a breaking change.

**Comment by santouits** [comment]
I also think that it should be named Open Simplex Noise because:

- That is it's correct name.
- It's a different algorithm.
- There is also a [wikipedia page](https://en.wikipedia.org/wiki/OpenSimplex_noise).
- It doesn't hide that there is an open alternative to the simplex noise, so it is important to the opensource culture.
- If in the future simplex noise patent expires, it can be added too without breaking old projects.

**Comment by JFonS** [comment]
I have no preference on the name other than SimplexNoise being shorter. Feel free to open a PR with the change or an appropriate issue to discuss this.

**Comment by Wircoark** [comment]
Are noise textures supposed to work within the Visual Shader Editor? I added a noise texture, fed it UV´s and put it in the Color output of a CanvasItem shader but it just shows up black. I tried right clicking the noise texture and tried to "edit" it so I could choose "SimplexNoise" like you would if you just add this texture to a Sprite but no settings are exposed. Figured I´d mention it here if it´s a bug, unless it´s something I´m missing on my part.

**Comment by JFonS** [comment]
@Wircoark It's a quirk of the visual shader editor. If you use the constant texture node and you create a new texture from the visual node itself then you have no way to modify it's parameters (gradient or curve textures have the same issue). 

I might try to fix it, but for the moment you can workaround it by creating a new NoiseTexture resource in memory (green plus icon in the inspector), saving it to a resource file and then loading it in the visual shader node. This way you will be able to open the resource's inspector through the file system dock and edit all of it's parameters.

**Comment by SethRzeszutek** [comment]
I am on the Alpha 3.1, when using the reference code from the docs I come up with OpenSimplexNoise not being declared in the current scope errors. Running the simple code in the Docs fail.
More information is on this is [https://godotengine.org/qa/35936/issue-using-opensimplex-godot-alpha-doesnt-register-module](here).

**Comment by Zylann** [comment]
@SethRzeszutek try 3.1 alpha2, it has simplex noise in it: https://downloads.tuxfamily.org/godotengine/3.1/alpha2/

**Comment by SethRzeszutek** [comment]
I am on the alpha, I downloaded it yesterday. Even the about section says GODOT 3.1.alpha, is there an alpha2?

**Comment by Zylann** [comment]
@SethRzeszutek it seems the editor doesn't display which alpha it is, but you should try the alpha2 link I provided above, this one definitely has OpenSimplexNoise.

**Comment by SethRzeszutek** [comment]
That link allowed for me to use openSimplex, I guess I somehow downloaded the wrong alpha? Thank you!
